File: stuff/tag/2021_03_16/continuumClasses/@dirichletClass/dirichletClass.m
classdef dirichletClass < handle
    properties (Constant)
        flagNewton = struct('initializeGlobalDofs',false,...
                            'initializeQV',false,...
                            'initializeShapeFunctions',false,...
                            'updateTimeDependentField',true,...
                            'updateContinuumFieldPreNewtonLoop',false,...
                            'callElements',false,...
                            'updateContinuumFieldNewtonLoop',false);
    end
    properties
        masterObject
        nodeList
        nodalDof = 1;       % 1,2,3         
        timeFunction = @(t,XYZ) 0;
        time = 0;
    end
    properties (Dependent = true)
        globalDirichletList
        qN1
    end

    %% constructor
    methods
        function obj = dirichletClass(dofObject)
            if nargin==0
                error('dofObject is required as input');
            end
            dofObject.listContinuumObjects{end+1} = obj;
        end
    end
    
    %% get und set methods
    methods
        function set.masterObject(obj,input)
            assert(isa(input,'solidClass')||isa(input,'solidThermoClass'),'masterObject must be compatibel to dirichlet class');
            obj.masterObject = input;
        end
        function set.timeFunction(obj,input)
            assert(isa(input,'function_handle'),'timeFunction must be of type function_handle');
            obj.timeFunction = input;
        end
        function set.nodeList(obj,input)
            checkIntegerMatrix(input);
            obj.nodeList = int32(input(:));
        end
        function set.nodalDof(obj,input)
            checkIntegerMatrix(input);
            obj.nodalDof = input;
        end
        function set.time(obj,t)
            assert(isnumeric(t),'time need to be of class double');
            obj.time = t; 
        end        
        function out = get.globalDirichletList(obj)   
            if ~isempty(obj.masterObject.globalNodesDof)
                list = obj.masterObject.globalNodesDof(obj.nodeList,obj.nodalDof);
                out = list(:);
            else
                error('globalNodesDof of masterObject is empty. Consider initialization via dofObject.initializeGlobalDofs!');
            end
        end
        function out = get.qN1(obj)
            if numel(obj.nodalDof)==1
                out = obj.timeFunction(obj.time,obj.masterObject.qR(obj.nodeList,obj.nodalDof));
            else
                out = obj.masterObject.qR(obj.nodeList,obj.nodalDof) + obj.timeFunction(obj.time);
            end
        end
    end
end

File: stuff/pyfemkit/continuumClasses/dirichletClass/dirichletClass.py
import numpy as np
import inspect
import sys
from commonFunctions.checkIntegerMatrix import checkIntegerMatrix

class dirichletClass(object):
    def __init__(self):
        self.flagNewton = {'initializeGlobalDofs': False,
                            'initializeQV': False,
                            'initializeShapeFunctions': False,
                            'updateTimeDependentField': True,
                            'updateContinuumFieldPreNewtonLoop': False,
                            'callElements': False,
                            'updateContinuumFieldNewtonLoop': False}
        #properties        
        self.masterObject
        self._nodeList = np.array([],dtype=int)
        self._nodalDof = 1   # 1,2,3  
        self._time = 0
        #dependent=true
        # self._globalDirichletList = np.array([],dtype=float) 
        # self.qN1 # = np.array([],dtype=float) 

    def dirichletClassfunc(self,dofObject):
        dofObject.listContinuumObjects.append(self) 
    
    def masterObject(self,input):
        if (inspect.isclass(input,'solidClass') or inspect.isclass(input,'solidThermoClass')):
            self.masterObject = input
        else:
            sys.exit('masterObject must be compatibel to dirichlet class!')
        
    def timeFunction(self,time,qR):
        # linear function
        # t = qR-time
        t = qR-0.5*time
        return t

    @property
    def nodeList(self):
        return self._nodeList
    @nodeList.setter
    def nodeList(self, input):
        # checkIntegerMatrix(input)
        self._nodeList = np.int32(input.reshape(-1,1)) 

    @property
    def nodalDof(self):
        return self._nodalDof
    @nodalDof.setter
    def nodalDof(self, input):
        # checkIntegerMatrix(input)solveDof
        self._nodalDof = input
            
    @property
    def time(self):
        return self._time
    @time.setter
    def time(self, input):
        assert isinstance(input, float),'time need to be class double!'
        self._time = input

    # @property
    def globalDirichletList(self):
        if not(self.masterObject.globalNodesDof.size == 0):
            lists = self.masterObject.globalNodesDof[self.nodeList,self.nodalDof-1]
            return lists.reshape(-1,1)
        else:
            return np.array([],dtype=int)
        
    # @property
    def qN1(self):
        if np.size(self.nodalDof)==1:
            return self.timeFunction(self.time,self.masterObject.qR[self.nodeList,self.nodalDof-1])
        else:
            return self.masterObject.qR[self.nodeList,self.nodalDof-1] + self.timeFunction[self.time]
